### DISCUSSION TITLE
Pathname.install doesn't work but manually copying files over works

### DISCUSSION THREAD

**Discussion by michaelmyc** [Homebrew/discussions#3151] [Tap maintenance and Homebrew development] [answered]
I'm trying to create a formulae for https://github.com/jasongin/nvs, and my install function is as follows. Running the `nvs-install.sh` script after `brew install --build-from-source nvs` results in the failing, but deleting the relevant files in prefix and pasting them from a freshly cloned repo works for me. Any idea what might have changed during the `install` method?

```rb
  def install
    (prefix/"nvs-install.sh").write <<~EOS
      # $NVS_HOME should be "$HOME/.nvs" by default to avoid user-installed nodes destroyed every update
      [ -z "$NVS_HOME" ] && export NVS_HOME="$HOME/.nvs"

      # nvs requires nvs.sh, defaults.json and the lib folder to work
      [ -e "$NVS_HOME" ] || mkdir -p "$NVS_HOME"
      [ -e "$NVS_HOME/nvs.sh" ] || ln -s #{prefix}/nvs.sh "$NVS_HOME/nvs.sh"
      [ -e "$NVS_HOME/defaults.json" ] || ln -s #{prefix}/defaults.json "$NVS_HOME/defaults.json"
      [ -e "$NVS_HOME/lib" ] || ln -s #{prefix}/lib "$NVS_HOME/lib"
      [ -e "$NVS_HOME/deps" ] || ln -s #{prefix}/deps "$NVS_HOME/deps"

      . $NVS_HOME/nvs.sh install
    EOS
    prefix.install "nvs.sh", "defaults.json", Dir["deps"], Dir["lib"]
    # system ". #{prefix}/nvs-install.sh"
  end
```

**Comment by SMillerDev** [comment]
You're not allowed to interact with anything outside the Homebrew prefix AFAIK in the install block.

**Reply by michaelmyc** [reply]
Yeah. I meant after brew install, the last line of the script throws errors, but when I delete the files in prefix and copy over the files from the manually cloned repo (without changing the symlinks), the last line would work fine.

**Comment by carlocab** [comment]
It's difficult to say without knowing what the error you're encountering is.

**Comment by michaelmyc** [comment, marked as answer]
Figured it out. It's an incompatibility with M1 in the release. Using the master branch resolved the issue. 

Quick question: is it acceptable to submit a package pointing to the master branch? Or is it more advised to use a Tap?

**Reply by gromgit** [reply]
> is it acceptable to submit a package pointing to the master branch?

Homebrew doesn't really care which branch you use, but it _does_ care that core formulae are built from stable tags. I assume you're really asking if a formula that _builds from HEAD by default_ is acceptable; the answer is a definite NO.

**Reply by mistydemeo** [reply]
If you know what the M1 fixes are, you could look into backporting those as patches to the stable release.

**Comment by michaelmyc** [comment]
Another question: does the Homebrew install process isolate the user's $HOME directory? The installer script will try to place files in the user's $HOME directory and write to the `.zshrc` file, but it seems like those files and changes would vanish. Any idea how I would do that? Or are they strictly forbidden and I have to ask the users to run the install script themselves?

**Reply by michaelmyc** [reply]
Also when first run, the executable will try to download a base version of Nodejs, and when I run the executable in the test function, it is refusing to run `mkdir`. How do I circumvent this?

**Reply by michaelmyc** [reply]
I don't think not testing the package is a good idea.

**Reply by gromgit** [reply]
> Another question: does the Homebrew install process isolate the user's $HOME directory? The installer script will try to place files in the user's $HOME directory and write to the `.zshrc` file, but it seems like those files and changes would vanish. Any idea how I would do that? Or are they strictly forbidden and I have to ask the users to run the install script themselves?

The Homebrew build process is sandboxed, so installing stuff outside `$HOMEBREW_PREFIX` is automatically denied. The best solution to your issue is to have the program run a script or something on startup, that checks for and automatically adds the necessary files and `.zshrc` additions. That by definition would work regardless of distribution mechanism _and_ user, and IMO is really how you should be architecting your project anyway.

**Reply by gromgit** [reply]
> Also when first run, the executable will try to download a base version of Nodejs, and when I run the executable in the test function, it is refusing to run `mkdir`. How do I circumvent this?

**Show** us the actual commands you're running and their output, and maybe someone can advise. **Tell** us a vague story about what's happening, which is what you just did, and no one's going to guess.

That said, if your formula requires Nodejs to run at all, it really should be declared as a dependency, rather than being vendored at runtime.

**Reply by gromgit** [reply]
> I don't think not testing the package is a good idea.

If you want to submit your formula to Homebrew core, "not testing" is not an option.

On your own tap, you get to choose.
